vllm/model_executor/layers/quantization/fp8.py

```diff
@@ -81,46 +98,163 @@ def create_weights(
+            # WEIGHT SCALE
+            weight_scale = Parameter(torch.empty(len(output_partition_sizes),
+                                                 dtype=torch.float32),
+                                     requires_grad=False)
+            layer.register_parameter("weight_scale", weight_scale)
+            set_weight_attrs(weight_scale, {
+                **extra_weight_attrs,
+                "shard_indexer":
+                self.scales_shard_indexer,
+            })
+
+            # ACTIVATION SCALE
+            if self.quant_config.activation_scheme == "static":
+                act_scale = Parameter(torch.empty(len(output_partition_sizes),
+                                                  dtype=torch.float32),
+                                      requires_grad=False)
+                layer.register_parameter("act_scale", act_scale)
+                set_weight_attrs(
+                    act_scale, {
+                        **extra_weight_attrs,
+                        "shard_indexer":
+                        self.scales_shard_indexer,
+                    })
+
+    def scales_shard_indexer(
+            self, param: torch.Tensor, loaded_weight: torch.Tensor,
+            shard_id: Union[str, int]) -> Tuple[torch.Tensor, torch.Tensor]:
+        qkv_idxs = {"q": 0, "k": 1, "v": 2}
+
+        if isinstance(shard_id, int):
+            pass
+        elif isinstance(shard_id, str):
+            if shard_id not in qkv_idxs:
+                raise ValueError(f"Unknown shard_id: {shard_id}")
+            shard_id = qkv_idxs[shard_id]
+        else:
+            ValueError(f"Shard id must be int or str but got {type(shard_id)}")
+
+        return param[shard_id], loaded_weight
 
     def process_weights_after_loading(self, layer: Module) -> None:
-        # Although the quant_method is propagated to all layers,
-        # only linear layers invoke "create_weights". So we check
-        # whether "weight_scaling_facor" is registered to determine
-        # whether the layer is a linear layer that requires quantization.
-        if not hasattr(layer, "weight_scaling_factor"):
+        if (not hasattr(layer, "process_after_load")
+                or not layer.process_after_load):
             return
 
-        qweight, weight_scale = ops.scaled_fp8_quant(layer.weight)
-        # torch._scaled_mm requires column-major in the second
-        # input (weight), so we transpose the quantized weight.
-        layer.weight = Parameter(qweight.t(), requires_grad=False)
-        layer.weight_scaling_factor.data.copy_(weight_scale)
+        # If checkpoint is fp1616 (not serialized fp8), quantize the weights.
+        if not self.quant_config.is_serialized:
+            qweight, weight_scale = ops.scaled_fp8_quant(layer.weight,
+                                                         scale=None)
+            layer.weight = Parameter(qweight.t(), requires_grad=False)
+            layer.weight_scale = Parameter(weight_scale, requires_grad=False)
```

Comment:
Why don't we need to do `layer.register_parameter` here since we skip it for weight_scale earlier?

Reply:
The reason we need `register_parameter` during `create_weights()` is because we use `module.named_parameters()` in `LlamaForCausalLM.load_weights()`.

Since this happens after weight loading, we do not need to do it